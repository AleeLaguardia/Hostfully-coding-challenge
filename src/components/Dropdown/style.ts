import styled from "styled-components";

export const Container = styled.div<{ error: boolean; width?: string }>`
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: ${(props) => props.width ? props.width : '160px'};
  height: fit-content;
  background-color: ${(props) => props.theme.colors.white};
  border-radius: 5px;
  border: ${(props) => props.error ? `1px solid ${props.theme.colors.radicalRed}` : 'none'};
  margin-top: 10px;
`;

export const Item = styled.div<{ isOpen: boolean }>`
  display: flex;
  min-height: 40px;
  width: 100%;
  background-color: ${(props) => props.theme.colors.white};
  justify-content: space-between;
  align-items: center;
  cursor: pointer;
  padding: 0 10px;

  &:hover {
    background-color: ${(props) => props.isOpen ? props.theme.colors.galery : 'transparent'};
  }

  span {
    color: ${(props) => props.theme.colors.purple};
    font-family: ${(props) => props.theme.fonts.poppins};
    font-size: ${(props) => props.theme.sizes.xs};
  }
`;

export const Icon = styled.img<{ isDropdownOpen: boolean }>`
  height: 20px;
  width: 20px;
  transition: 0.3s ease-in-out;
  transform: rotate(${(props) => props.isDropdownOpen ? '180deg' : 0});
`;

export const OptionsContainer = styled.div<{ isDropdownOpen: boolean; width?: string }>`
  @keyframes changeOpacityOpen {
    0% { opacity: 0 },
    100% { opacity: 1 },
  };

  position: absolute;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: ${(props) => props.width ? props.width : '160px'};
  max-height: 170px;
  background-color: ${(props) => props.theme.colors.white};
  border-bottom-right-radius: 5px;
  border-bottom-left-radius: 5px;
  border: none;
  overflow-y: auto;
  margin-top: 40px;
  animation: changeOpacityOpen 0.3s;
  box-shadow: 0px 10px 5px 0px rgba(0,0,0,0.2);
`;