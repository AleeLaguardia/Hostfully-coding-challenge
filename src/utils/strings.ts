export const PATH = {
  HOME: '/',
  BOOKING: '/booking',
  RESERVATION: '/reservation',
};

export const HOME = {
  TITLE_FORM: "Book your stay right now!",
  TITLE_PIC: "Book Your Stay",
  ADULTS: "How many adults",
  CHILDREN: "How many children",
}

export const BOOKING = {
  NO_DATA: "No hotels found in that area",
  ALSO_TRY: "But you can also try:"
};