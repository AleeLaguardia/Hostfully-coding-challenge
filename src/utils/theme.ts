import { DefaultTheme } from "styled-components";

import {
  white,
  alabaster,
  emperor,
  galery,
  gray,
  black,
  purple,
  purpleHover,
  darkPurple,
  radicalRed,
  lightPurple,
  silverChalice,
  alto,
  lynch,
  selectiveYellow,
  chelseaGem,
} from './styling/colors';
import {
  poppins,
  playwrite,
} from './styling/fonts';
import {
  xs,
  xxs,
  xxxs,
  s,
  m,
  l,
  xl,
  xxxl,
} from './styling/sizes';

export const theme: DefaultTheme = {
  colors: {
    white,
    alabaster,
    emperor,
    galery,
    gray,
    black,
    purple,
    purpleHover,
    darkPurple,
    radicalRed,
    lightPurple,
    silverChalice,
    alto,
    lynch,
    selectiveYellow,
    chelseaGem,
  },
  fonts: {
    poppins,
    playwrite,
  },
  sizes: {
    xs,
    xxs,
    xxxs,
    s,
    m,
    l,
    xl,
    xxxl,
  },
}