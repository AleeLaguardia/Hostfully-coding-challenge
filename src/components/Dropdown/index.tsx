import React from "react";
import ArrowIcon from '../../assets/icons/arrow.svg';
import { Container, Icon, Item, OptionsContainer } from "./style";

const example = ['Place 1', 'Place 2', 'Place 3', 'Place 4', 'Place 5', 'Place 6'];

interface Props {
  isDropdownOpen: boolean;
  optionSelected: string;
  options?: string[];
  setIsDropdownOpen: React.Dispatch<React.SetStateAction<boolean>>;
  setOptionSelected: React.Dispatch<React.SetStateAction<string>>;
  placeholder: string;
  'data-testid'?: string;
  error: boolean;
  width?: string;
}

const Dropdown: React.FC<Props> = ({
  options = example,
  isDropdownOpen,
  optionSelected,
  setIsDropdownOpen,
  setOptionSelected,
  placeholder,
  error,
  'data-testid': testId,
  width,
}) => {
  const handleOptionSelected = (option: string) => {
    setOptionSelected(option);
  }

  return (
    <Container width={width} error={error} onClick={() => setIsDropdownOpen((prevState) => !prevState)} data-testid={testId}>
      <Item isOpen={false}>
        <span>{optionSelected || placeholder}</span>
        <Icon src={ArrowIcon} alt="ArrowIcon" isDropdownOpen={isDropdownOpen} />
      </Item>
      {isDropdownOpen && (
        <OptionsContainer width={width} isDropdownOpen={isDropdownOpen}>
          {options.map((i, index) => (
            <Item isOpen key={index} onClick={() => handleOptionSelected(i)}>
              <span>{i}</span>
            </Item>
          ))}
        </OptionsContainer>
      )}
    </Container>
  );
};

export default Dropdown;
