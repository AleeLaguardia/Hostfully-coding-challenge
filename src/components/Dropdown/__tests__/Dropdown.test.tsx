import React from 'react';
import { fireEvent, getByText, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Dropdown from '..';
import { theme } from '../../../utils/theme';
import { ThemeProvider } from 'styled-components';

describe('Dropdown Component', () => {
  const example = ['Place 1', 'Place 2', 'Place 3', 'Place 4', 'Place 5', 'Place 6'];

  const isDropdownOpenMock = false;
    const optionSelectedMock = 'test'
    const setIsDropdownOpen = jest.fn();
    const setOptionSelectedMock = jest.fn();

  it('renders Dropdown with the default style', () => {
    const { asFragment } = render(
      <ThemeProvider theme={theme}>
        <Dropdown
          data-testid='testid'
          options={example}
          isDropdownOpen={isDropdownOpenMock}
          optionSelected={optionSelectedMock}
          setIsDropdownOpen={setIsDropdownOpen}
          setOptionSelected={setOptionSelectedMock}
          placeholder='test'
        />
      </ThemeProvider>
    );

    expect(asFragment).toMatchSnapshot();
  });

  it('Opens dropdown when clicked', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Dropdown
          data-testid='testid'
          placeholder='test'
          options={example}
          isDropdownOpen={isDropdownOpenMock}
          optionSelected={optionSelectedMock}
          setIsDropdownOpen={setIsDropdownOpen}
          setOptionSelected={setOptionSelectedMock}
        />
      </ThemeProvider>
    );

    const dropdown = getByText(container, 'test');
    fireEvent.click(dropdown);
    expect(setIsDropdownOpen).toHaveBeenCalled();
    expect(isDropdownOpenMock).toBeTruthy;
  })

  it('Selects an option when clicking an item', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Dropdown
          data-testid='testid'
          placeholder='test'
          options={example}
          isDropdownOpen={true}
          optionSelected={example[0]}
          setIsDropdownOpen={setIsDropdownOpen}
          setOptionSelected={setOptionSelectedMock}
        />
      </ThemeProvider>
    );

    const dropdownOption = getByText(container, 'Place 2');
    fireEvent.click(dropdownOption);
    expect(setOptionSelectedMock).toHaveBeenCalledWith('Place 2');
    expect(setIsDropdownOpen).toHaveBeenCalled();
  })
});
