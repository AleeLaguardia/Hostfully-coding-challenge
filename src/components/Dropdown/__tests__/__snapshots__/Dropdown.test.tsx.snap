// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component renders Dropdown with the default style 1`] = `[Function]`;
